### DISCUSSION TITLE
Confusion regarding Modifier Keys mappings (eg. Control >> Control)

### DISCUSSION THREAD

**Discussion by enesbala5** [houmain/keymapper#174] [Q&A] [answered]
Hello,

I have been using `keymapperd` for a while now, and it's been great. However I noticed that I could no longer Ctrl + Click on a link to open it in a new tab.

I went to check the documentation here and learned about this type of declaration:

```
# Pass Modifier Keys through

Control >> Control
Shift >> Shift
Alt >> Alt
``` 

After adding this to the top of the file, my keybindings stopped working mostly, as they utilize Control, Shift, etc.. However my Ctrl + Click started working again.

I then moved the code snippet to the bottom and keybindings worked, but the Ctrl + Click didn't.

Can anyone please suggest a solution to this situation?

---

_Here is my full config for reference:_

```
# ------------------------------------------------------------------------
# Aliases
# ------------------------------------------------------------------------
Alt   = AltLeft
AltGr = AltRight
Win   = Meta
Left   = ArrowLeft
Right   = ArrowRight

# ------------------------------------------------------------------------
# Pass Modifier Keys through
# ------------------------------------------------------------------------
Control >> Control
Shift >> Shift
Alt >> Alt


# ------------------------------------------------------------------------
# Global mappings
# ------------------------------------------------------------------------
PageUp >> AudioVolumeUp
PageDown >> AudioVolumeDown
Home >> MediaPlayPause
Insert >> Meta{A}
End >> F11
MediaTrackPrevious >> NumLock
MediaTrackNext >> NumLock
# (Control Alt){P} >> (Control Alt){F6}

[system = "Linux"]
(Control Alt){P} >> (Control Shift){F6}


# ------------------------------------------------------------------------
# Application-specific mappings
# ------------------------------------------------------------------------

# Zen
# ---------
[system = "Linux", class = "zen-alpha"]
[system = "Windows", path = "zen.exe"]

(Control Shift){P} >> Control{L}
Control{P} >> Control{T}
Control{D} >> Alt{P}
(Control Alt){Left} >> (Control Alt Shift){U}
(Control Alt){Right} >> (Control Alt Shift){U}

# Obsidian
# ---------
# [system = "Linux", class = "obsidian"]

# For Future Focus Mode
# (Control K) Z >> (Control Alt Shift){Z}
```

**Comment by houmain** [comment, marked as answer]
Hi, you likely have to undo the modifiers which were forwarded before and interfere with the shortcuts, like:
```
(Control Alt){P} >> !Alt (Control Shift){F6}
(Control Shift){P} >> !Shift Control{L}
Control{D} >> !Control Alt{P}
```
Are these the ones which stopped working?

**Reply by enesbala5** [reply]
This worked. Very odd solution though - I think it should be mentioned somewhere. I also don't feel like my use case is rare - so this would probably help many like me to explore keymapper.

**Reply by houmain** [reply]
Yes, right. It also has been bugging me for a while now. [Here](#167) I have been writing about a possible solution, which I implemented now. In 4.8.1 there is a new directive [@forward-modifiers](https://github.com/houmain/keymapper?tab=readme-ov-file#directives) to try out.

**Reply by ristomatti** [reply]
@enesbala5 It's actually pretty useful how it behaves currently. For instance, I have a config like this:

```starlark
AltRight{A} >> AltRight{Q} # 'ä'
AltRight{O} >> AltRight{P} # 'ö'
```

This also implies that `Shift{AltRight{A}}` will output `Ä` as Keymapper doesn't automatically release keys pressed at the same time.

**Reply by ristomatti** [reply]
@houmain This tripped me a couple of days as well. I was pulling my hair off trying to figure why i3wm mapping didn't work as I wanted. I use `Meta{Digit}` to switch workspaces and added an alternative mapping, e.g `Meta{W} >> !W Meta{0}` which seemed to work fine, but then I noticed `Meta` gets released after it triggers and I couldn't then follow up with `IJKL` to focus window tiles directionally (while still holding `Meta`).

The issue turned out to be that I was forwarding the mods only on the first stage and not the second one. I see the new directive should do this automatically for all stages, nice!
